analytic_engine/src/sst/parquet/lazy_row_filter.rs

```diff
@@ -0,0 +1,445 @@
+/// when unselected.
+///
+/// For simplicity and consideration of parquet's related
+/// implementation (pushed filter must be highly selective, otherwise it will
+/// even add up the cost of decoding...), the selected filter should satisfy:
+///
+///   - Has just one column(and not nested).
+///   - Highly selective(now just `=`, `in`).
+///
+/// Detailed selection logic can be seen in [FilterStats].
+struct FilterCandidatesSelector<'a> {
+    exprs: &'a [Expr],
+    schema: &'a Schema,
+    meta_data: &'a ParquetMetaData,
+}
+
+impl<'a> FilterCandidatesSelector<'a> {
+    fn new(exprs: &'a [Expr], schema: &'a Schema, meta_data: &'a ParquetMetaData) -> Self {
+        Self {
+            exprs,
+            schema,
+            meta_data,
+        }
+    }
+
+    fn select(self) -> Result<Vec<FilterCandidate>> {
+        self.exprs
+            .iter()
+            .filter_map(|expr| {
+                FilterCandidate::try_new(expr, self.schema, self.meta_data).transpose()
+            })
+            .collect()
+    }
+}
+
+/// Sorter for filter candidates
+// TODO: make it a trait if extensible required in later.
+struct FilterCandidatesSorter(Vec<FilterCandidate>);
+
+impl FilterCandidatesSorter {
+    fn sort(mut self) -> Vec<FilterCandidate> {
+        self.0
+            .sort_unstable_by_key(|candidate| candidate.required_bytes);
+        self.0
+    }
+}
+
+/// Calculate the total compressed size of all `Column's required for
+/// predicate `Expr`. This should represent the total amount of file IO
+/// required to evaluate the predicate.
+fn size_of_columns(columns: &BTreeSet<usize>, metadata: &ParquetMetaData) -> usize {
+    let mut total_size = 0;
+    let row_groups = metadata.row_groups();
+    for idx in columns {
+        for rg in row_groups.iter() {
+            total_size += rg.column(*idx).compressed_size() as usize;
+        }
+    }
+
+    total_size
+}
+
+/// Filter stats used for selecting filter candidate
+struct FilterStats {
+    required_column_indices: BTreeSet<usize>,
+    non_primitive_columns: bool,
+    highly_selective: bool,
+}
+
+impl FilterStats {
+    fn new() -> Self {
+        Self {
+            required_column_indices: BTreeSet::new(),
+            non_primitive_columns: true,
+            highly_selective: false,
+        }
+    }
+
+    fn selected(&self) -> bool {
+        self.required_column_indices.len() == 1
+            && self.non_primitive_columns
+            && self.highly_selective
+    }
+}
+
+struct FilterStatsBuilder<'a> {
+    expr: &'a Expr,
+    schema: &'a Schema,
+    stats: FilterStats,
+}
+
+impl<'a> FilterStatsBuilder<'a> {
+    fn new(expr: &'a Expr, schema: &'a Schema) -> Self {
+        Self {
+            expr,
+            schema,
+            stats: FilterStats::new(),
+        }
+    }
+
+    fn build(mut self) -> Result<FilterStats> {
+        self.expr.visit(&mut self).context(DataFusionError)?;
+
+        self.stats.highly_selective = match self.expr {
```

Comment:
Maybe whether a filter is `highly_selective` depends on not only the filter form but also the number of the unique values of the filtered column. If the column has only a few values, the filters applied to it may be badly selective.

And is there any information describing the UV of the columns in the metadata?